pypesto/objective/roadrunner/utils.py

```diff
@@ -0,0 +1,469 @@
+
+    def get_observable_ids(self):
+        """
+        Get the observable ids of the measurement data.
+
+        Returns
+        -------
+        observable_ids:
+            Observable ids of the measurement data.
+        """
+        return self.observable_ids
+
+    def sanity_check(self):
+        """Perform a sanity check of the data."""
+        if not np.allclose(self.measurements[:, 0], self.timepoints):
+            raise ValueError(
+                "Timepoints do not match the first column of measurements."
+            )
+        if self.measurements.shape[1] != len(self.observable_ids) + 1:
+            raise ValueError(
+                "Number of columns in measurements does not match number of "
+                "observable ids + time."
+            )
+        # check that the noise distributions and noise formulae have the
+        # same length as the number of observables
+        if len(self.noise_distributions) != len(self.observable_ids):
+            raise ValueError(
+                "Number of noise distributions does not match number of "
+                "observable ids."
+            )
+        if len(self.noise_formulae) != len(self.observable_ids):
+            raise ValueError(
+                "Number of noise formulae does not match number of "
+                "observable ids."
+            )
+
+    @staticmethod
+    def from_petab_problem(petab_problem: petab.Problem) -> list[ExpData]:
+        """
+        Create a list of ExpData object from a petab problem.
+
+        Parameters
+        ----------
+        petab_problem:
+            PEtab problem.
+        """
+        # extract all condition ids from measurement data
+        condition_ids = list(
+            petab_problem.measurement_df["simulationConditionId"].unique()
+        )
+        exp_datas = [
+            ExpData.from_petab_single_condition(
+                condition_id=condition_id, petab_problem=petab_problem
+            )
+            for condition_id in condition_ids
+        ]
+        return exp_datas
+
+    @staticmethod
+    def from_petab_single_condition(
+        condition_id: str, petab_problem: petab.Problem
+    ) -> ExpData:
+        """
+        Create an ExpData object from a single condition of a petab problem.
+
+        Parameters
+        ----------
+        condition_id:
+            Identifier of the condition.
+        petab_problem:
+            PEtab problem.
+        """
+        # extract measurement data for a single condition
+        measurement_df = petab_problem.measurement_df[
+            petab_problem.measurement_df[SIMULATION_CONDITION_ID]
+            == condition_id
+        ]
+        # turn measurement data into a numpy array
+        measurements, observale_ids = measurement_df_to_matrix(measurement_df)
+        # timepoints are the first column of the measurements array
+        timepoints = measurements[:, 0]
+        # TODO: Needs to be revised when roadrunner allows for duplicate time
```

Comment:
The time has come: https://github.com/sys-bio/roadrunner/releases/tag/v2.6.0

Reply:
yes I know :D thanks :)

Reply:
integrated 👍🏼